### PR TITLE
enh: support assigning array dimension via module variables

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes: #3952.

### CODE REVIEW

**Comment by Pranavchiku** [comment]
This pr transforms:

```fortran
module xx
integer :: nx = 4
end module xx

subroutine a(cs)
use xx
real, dimension(nx), intent(in) :: cs
print *, cs
end subroutine

program main
real, dimension(4) :: cs
cs = [1.0, 2.0, 3.0, 4.0]
call a(cs)
end program
```

TO

```
module xx
implicit none
integer(4), save :: nx = 4
end module xx

program main
implicit none
real(4), dimension(4) :: cs
cs = [1.000000, 2.000000, 3.000000, 4.000000]
call a(cs)

contains

pure integer(4) function __lcompilers_get_nx()
    __lcompilers_get_nx = nx
end function __lcompilers_get_nx

subroutine a(cs)
    real(4), dimension(__lcompilers_get_nx()), intent(in) :: cs
    interface
        pure integer(4) function __lcompilers_get_nx()
        end function __lcompilers_get_nx

    end interface
    print *, cs
end subroutine a

end program main
```

**Comment by Pranavchiku** [comment]
```fortran
module xx
integer :: nx = 4
integer :: ny = 4
integer :: nz = 4

contains

subroutine a(cs)
    real, dimension(nx, ny, nz), intent(in) :: cs
    print *, cs
end subroutine

end module xx
```

This still do not work, I know how to fix this, will push it soon.

**Review by certik — APPROVED** [review]
I think this looks great, thanks for fixing it. The nbody test needs to be uncommented before we can merge it.

File: integration_tests/CMakeLists.txt [hunk 1]
```diff
@@ -1114,7 +1117,7 @@ RUN(NAME nested_vars1 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc wasm c)
 RUN(NAME nested_vars2 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc wasm c)
 RUN(NAME nested_vars3 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc wasm c)
 
-RUN(NAME nbody LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
+# RUN(NAME nbody LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
```
[the remark this answers is not in the document]
Comment (Pranavchiku) [reply integration_tests/CMakeLists.txt:1120]: Aah, I did uncomment it, it works fine, I’ll push it.

**Pull request merged**